### ISSUE TITLE
Using on exit handlers with semaphores leads to workflow operation error on retry

### ISSUE THREAD

**Issue by louisnow** [argoproj/argo-workflows#7934]
#### What happened
Using on exit handlers with semaphores leads to workflow operation error on retry. Similar to https://github.com/argoproj/argo-workflows/issues/7888 from the logs
`invalid memory address or nil pointer dereference`

https://user-images.githubusercontent.com/6024271/154860669-4d92e078-2f0c-43f1-aa3a-5a0ebe546e56.mov

#### What you expected to happen?
It should instead work properly similar to when you remove the semaphore

https://user-images.githubusercontent.com/6024271/154861338-48268fc9-1f13-4d1a-b64b-4e5eaeda121d.mov

What executor are you using? PNS/Emissary
Emissary

## Diagnostics

Paste the smallest workflow that reproduces the bug. We must be able to run the workflow.
Workflow Configmap
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: test-retry-config
data:
  hello: "2"
```

Workflow template
```yaml
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: retry
spec:
  entrypoint: main
  onExit: exit-dag
  templates:
    - name: main
      inputs: 
        parameters:
          - name: numbers  
            value: |
              ["one", "two", "three"]
      dag:
        tasks:
          - name: whalesay-before-process
            template: whalesay
            arguments:
              parameters:
                - name: message
                  value: "HELLO ALL"

          - name: process
            template: process
            withParam: "{{inputs.parameters.numbers}}"
            dependencies:
              - whalesay-before-process
            arguments:
              parameters:
                - name: message
                  value: "{{item}}"

          - name: whalesay-after-process
            template: whalesay
            dependencies:
              - process
            arguments:
              parameters:
                - name: message
                  value: "Root: {{tasks.process.outputs.parameters.output}}"           

    - name: process
      # this fixes the behaviour
      retryStrategy:
        limit: "0"
      inputs:
        parameters:
        - name: message
      outputs:
        parameters:
          - name: output
            valueFrom:
              parameter: "{{tasks.step2.outputs.parameters.output}}"
      dag: 
        tasks:
          - name: step0
            template: whalesay
            arguments:
              parameters:
                - name: message
                  value: "{{inputs.parameters.message}}"

          - name: step1
            template: whalesay
            dependencies:
              - step0
            arguments:
              parameters:
                - name: message
                  value: "{{inputs.parameters.message}}"
                - name: always-pass
                  value: "false"

          - name: step2
            template: whalesay
            dependencies:
              - step1
            arguments:
              parameters:
                - name: message
                  value: "{{tasks.step1.outputs.parameters.output}}"

    - name: whalesay
      synchronization:
        semaphore:
          configMapKeyRef:
            name: test-retry-config
            key: hello
      inputs:
        parameters:
        - name: message
        - name: always-pass
          value: "true"
      outputs:
        parameters:
          - name: output
            valueFrom:
              path: /tmp/output.txt
      script:
        command: [ python ]
        image: python:alpine
        imagePullPolicy: IfNotPresent
        source: |
          import random
          import os
          msg = '{{inputs.parameters.message}}'
          print(msg)

          if '{{inputs.parameters.always-pass}}' == 'true':
            a = -1
          else:
            a = random.randint(1,10)          

          if a > 0 and a <= 5: 
            raise Exception(msg)

          with open('/tmp/output.txt', 'w') as o:
            o.write('random' + msg + ' : ' +str(a))

    - name: exit-dag
      dag: 
        tasks:
          - name: exit-msg
            template: exit-msg

    - name: exit-msg
      container:
        image: docker/whalesay
        command: [cowsay]
        args: ["goodbye world"]
```

What executor are you running? PNS/Emissary
Emissary

```bash
# Logs from the workflow controller:
kubectl logs -n argo deploy/workflow-controller | grep ${workflow} 

# If the workflow's pods have not been created, you can skip the rest of the diagnostics.

# The workflow's pods that are problematic:
kubectl get pod -o yaml -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded

# Logs from in your workflow's wait container, something like:
kubectl logs -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```
Workflow controller logs
```
time="2022-02-20T19:27:49.631Z" level=info msg="archiving workflow" namespace=default uid=747dc8ac-17f8-4a03-9c77-26fca386fcab workflow=retry-twskk
time="2022-02-20T19:27:49.638Z" level=info msg="Queueing Failed workflow default/retry-twskk for delete in 24h0m0s"
time="2022-02-20T19:27:49.638Z" level=info msg="Queueing Failed workflow default/retry-twskk for delete in 24h0m0s"
time="2022-02-20T19:28:18.405Z" level=info msg="Processing workflow" namespace=default workflow=retry-twskk
time="2022-02-20T19:28:18.406Z" level=error msg="Recovered from panic" namespace=default r="runtime error: invalid memory address or nil pointer dereference" stack="goroutine 2737 [running]:\nruntime/debug.Stack(0xc0f4f2b0a9, 0x1f22520, 0x331a940)\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0x9f\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate.func2(0xc000776cc0, 0x24f2de8, 0xc00005c018)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:192 +0xd1\npanic(0x1f22520, 0x331a940)\n\t/usr/local/go/src/runtime/panic.go:971 +0x499\ngithub.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1.(*SynchronizationStatus).GetStatus(...)\n\t/go/src/github.com/argoproj/argo-workflows/pkg/apis/workflow/v1alpha1/workflow_types.go:3116\ngithub.com/argoproj/argo-workflows/v3/workflow/sync.(*Manager).Release(0xc000446300, 0xc003505800, 0xc00016b4d0, 0x15, 0xc001ebdd70)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/sync/sync_manager.go:195 +0x373\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc000776cc0, 0x24f2de8, 0xc00005c018, 0xc004fd5570, 0xc000bbab61, 0x17)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:358 +0x1fe5\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc000776cc0, 0x24f2de8, 0xc00005c018, 0xc004fd5570, 0xc00224ca81, 0x7)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:408 +0x4a5\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc000776cc0, 0x24f2de8, 0xc00005c018, 0xc004fd5570, 0xc006b1ed38, 0x16)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:408 +0x4a5\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAG(0xc000776cc0, 0x24f2de8, 0xc00005c018, 0xc0023aa600, 0xb, 0xc0014c1340, 0xc0043bca48, 0x6, 0xc0024b7200, 0x24dd2b0, ...)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:244 +0x309\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeTemplate(0xc000776cc0, 0x24f2de8, 0xc00005c018, 0xc0023aa600, 0xb, 0x24dd2b0, 0xc000776d80, 0xc0014c1300, 0x0, 0x0, ...)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1807 +0x16bc\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate(0xc000776cc0, 0x24f2de8, 0xc00005c018)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:344 +0xf3e\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).processNextItem(0xc000101000, 0x24f2de8, 0xc00005c018, 0x0)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:770 +0x945\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).runWorker(0xc000101000)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:692 +0x9b\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc001616b20)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:155 +0x5f\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc001616b20, 0x24ab840, 0xc002f5b920, 0x7469746e45646501, 0xc000f006c0)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:156 +0x9b\nk8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc001616b20, 0x3b9aca00, 0x0, 0x6c203d2073656901, 0xc000f006c0)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:133 +0x98\nk8s.io/apimachinery/pkg/util/wait.Until(0xc001616b20, 0x3b9aca00, 0xc000f006c0)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:90 +0x4d\ncreated by github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).startLeading\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:309 +0x48a\n" workflow=retry-twskk
time="2022-02-20T19:28:18.406Z" level=info msg="Updated phase Running -> Error" namespace=default workflow=retry-twskk
time="2022-02-20T19:28:18.406Z" level=info msg="Updated message  -> runtime error: invalid memory address or nil pointer dereference" namespace=default workflow=retry-twskk
time="2022-02-20T19:28:18.406Z" level=info msg="Marking workflow completed" namespace=default workflow=retry-twskk
time="2022-02-20T19:28:18.406Z" level=info msg="Marking workflow as pending archiving" namespace=default workflow=retry-twskk
time="2022-02-20T19:28:18.406Z" level=info msg="Checking daemoned children of " namespace=default workflow=retry-twskk
time="2022-02-20T19:28:18.450Z" level=info msg="Workflow update successful" namespace=default phase=Error resourceVersion=2606764 workflow=retry-twskk
time="2022-02-20T19:28:18.458Z" level=info msg="archiving workflow" namespace=default uid=747dc8ac-17f8-4a03-9c77-26fca386fcab workflow=retry-twskk
time="2022-02-20T19:28:18.662Z" level=error msg="failed to archive workflow" error="failed to archive workflow: pq: duplicate key value violates unique constraint \"argo_archived_workflows_pkey\"" key=default/retry-twskk
time="2022-02-20T19:28:18.663Z" level=info msg="archiving workflow" namespace=default uid=747dc8ac-17f8-4a03-9c77-26fca386fcab workflow=retry-twskk
time="2022-02-20T19:28:18.697Z" level=info msg="Queueing Error workflow default/retry-twskk for delete in 24h0m0s"
time="2022-02-20T19:28:18.699Z" level=info msg="Queueing Error workflow default/retry-twskk for delete in 24h0m0s"
time="2022-02-20T19:41:37.354Z" level=info msg="Queueing Error workflow default/retry-twskk for delete in 23h46m41s"
```
---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by louisnow** [comment]
I've noticed it happening even without the exit handler in more complicated DAGs where we use semaphores.
Using parallelism could be a workaround but we're [facing issues](https://github.com/argoproj/argo-workflows/issues/7936) with that approach as well.

**Comment by sarabala1979** [comment]
@louisnow Issue fixed and release in 3.3.0-rc8. Can you test it?

**Comment by louisnow** [comment]
Is there any chance of back porting the fix to the latest stable version(3.2.x) that is released @sarabala1979 ?
